Update filename regex to support non-ascii letters

sunovivid opened the pull request:
I changed regex of filename to support non-ascii (unicode) letters.

Pushpavel commented:
Hi sunovivid, Thanks for the pull request. Actually, file creation for java is handled a bit different where this filename patching is done for the sub-directory created exclusively for that problem. ```JavaFileGenerator``` is the class in the same folder as ```DefaultFileGenerator``` class. Also, you can remove the whitespace and hyphen filtering too.

sunovivid commented:
Thank you for your answer!

I didn't know that I missed `JavaFileGenerator`. Thanks for notifying me.

I added a new commit.

First, I added common `getValidFileName` function for both `DefaultFileGenerator` and `JavaFileGenerator`. 

Second, I changed the special character filtering based on [this answer](https://stackoverflow.com/a/47455094/19194418). 

But blacklist-filtering has some potential dangers as I think (like missing reserved special characters or OS-specific reserved words), so I added a validation on filename by delegating it to OS. If the validation failed, it returns valid filename based on whitelist-filtering.

And I found that whitelist-filtering can have a danger too. For example, whitespaces after the filename cause `InvalidPathException` in the function `Path.gets(filename)`. So I added trim() but still there can exist some dangers. So if the whitelist-filtering also fails, it uses the original whitespaces and hyphen filtering. I think it is much safer.

Please tell me if the code is too complicated or not readable.

Pushpavel commented:
Hi sunovivid, Thanks for the extensive research, the code looks great.
